File: runme/mcgen/local/datan/solsrc/s2lsq.f
      PROGRAM S2LSQ
      IMPLICIT DOUBLE PRECISION(A-H,O-Z)
      PARAMETER(MAXNR=1,MAXN=100)
      DIMENSION T(MAXN),DELTAY(MAXN),CX(MAXNR,MAXNR),A(MAXN,MAXNR)
      DIMENSION C(MAXN),Y(MAXN),X(MAXNR),SCRAT(MAXNR,MAXNR)
      CHARACTER*75 TX,TY,CAPT
      DIMENSION DATSX(MAXN),DATCOV(MAXN)
      DIMENSION XPL(200),YPL(200)
      LOGICAL OK
      DATA NR /1/
C identify program to user
      WRITE(*,*)' Program S2LSQ generates data'
      WRITE(*,*)' and fits a power law (linear case)'
      WRITE(*,*)' '
C ask for input
      WRITE(*,*)' Enter number n of data points (2 <= n <= 100))'
      WRITE(*,*)' >'
      READ(*,*)N
      WRITE(*,*)' Enter first value T0 of controlled variable'
      WRITE(*,*)' >'
      READ(*,*)T0
      WRITE(*,*)' Enter step width DELTAT of controlled variable'
      WRITE(*,*)' >'
      READ(*,*)DELTAT
      WRITE(*,*)' Enter coefficient X'
      WRITE(*,*)' >'
      READ(*,*)X(1)
      WRITE(*,*)' Enter exponent W'
      WRITE(*,*)' >'
      READ(*,*)W
      WRITE(*,*)' Enter size of measurement errors'
      WRITE(*,*)' >'
      READ(*,*)SIGMA
      CALL GRNBWS()
      WRITE(*,*)' Please, enter number of workstation:'
      WRITE(*,*)' >'
      READ(*,*)NWS
C generate data points corresponding to power law
      CALL RNSTNR(DELTAY,N)
      DO 10 I=1,N
        T(I)=T0+DBLE(I-1)*DELTAT
        Y(I)=X(1)*(T(I)**W)
        Y(I)=Y(I)+DELTAY(I)*SIGMA
        DELTAY(I)=SIGMA
10    CONTINUE
C set up A and C
      DO 15 I=1,N
        C(I)=-Y(I)
        A(I,1)=-(T(I)**W)
15    CONTINUE
C perform fit
      CALL LSQLIN(T,C,DELTAY,N,NR,X,CX,R,A,SCRAT,OK)
C compute chi**2 probability
      IF(NR.EQ.N) THEN
        P=0.D0
      ELSE
        P=1.D0-SCCHI2(R,N-NR)
      END IF
C prepare graphics
      TX='t'
      TY='y'
      LTX=1
      LTY=1
      CAPT='x#=******, &D@x=******, M=******, P=******'
      WRITE(CAPT(4:9),'(F6.2)')X(1)
      WRITE(CAPT(17:22),'(F6.2)')SQRT(CX(1,1))
      WRITE(CAPT(27:32),'(F6.2)')R
      WRITE(CAPT(37:42),'(F6.4)')P
      LCAPT=42
C curve corresponding to solution
      NPL=200
CCCC          DT=(DBLE(N-1)/DBLE(NPL-1))*DELTAT
      DT=(DBLE(N)/DBLE(NPL-1))*DELTAT
      DO 20 IPL=1,NPL
        XPL(IPL)=T0+DT*DBLE(IPL)
        YPL(IPL)=X(1)*(XPL(IPL)**W)
20    CONTINUE
C prepare data points for graphical presentation
      DO 30 I=1,N
        DATSX(I)=0.D0
        DATCOV(I)=0.D0
30    CONTINUE
C graphical output
      CALL GRDTCV(XPL,YPL,NPL,1,.5D0,T,Y,DATSX,DELTAY,DATCOV,N,
     +TX,LTX,TY,LTY,CAPT,LCAPT,NWS)
      END

File: runme/mcgen/local/datan/solsrc/s2lsq.c
#include <stdio.h>
#include <math.h>
#include <string.h>
#include "datsrc.h"
#include "grsrc.h"

/* Table of constant values */

static integer c__1 = 1;
static double c_dp5 = .5;

main(void)
{
    /* Local variables */
    static integer i, n, ipl, npl, lcapt, ltx, lty, nws, nr = 1;
    static double a[100] /* was [100][1] */, c[100], cx[1] /* was [1][1] */,
                  deltat, datcov[100], deltay[100], dt, p, r, t[100], w, x[1],
                  y[100], sigma, scrat[1] /* was [1][1] */, datsx[100], t0,
                  xpl[200], ypl[200];
    static char capt[75], tx[75], ty[75];
    static logical ok;
    /* System generated locals */
    integer i__1;
/* identify program to user */
    printf("%s\n", "Program S2LSQ generates data");
    printf("%s\n\n", "and fits a power law (linear case).");
/* ask for input */
    printf("%s\n", "Enter number n of data points (2 <= n <= 100)):");
    printf("%s", "> ");
    scanf("%li", &n);
    printf("%s\n", "Enter first value T0 of controlled variable:");
    printf("%s", "> ");
    scanf("%lf", &t0);
    printf("%s\n", "Enter step width DELTAT of controlled variable:");
    printf("%s", "> ");
    scanf("%lf", &deltat);
    printf("%s\n", "Enter coefficient X:");
    printf("%s", "> ");
    scanf("%lf", &x[0]);
    printf("%s\n", "Enter exponent W:");
    printf("%s", "> ");
    scanf("%lf", &w);
    printf("%s\n", "Enter size of measurement errors:");
    printf("%s", "> ");
    scanf("%lf", &sigma);
    grnbws_();
    printf("%s\n", "Please, enter number of workstation:");
    printf("%s", "> ");
    scanf("%li", &nws);
    (void)getchar();
/* generate data points corresponding to power law */
    rnstnr_(deltay, &n);
    i__1 = n;
    for (i = 1; i <= i__1; ++i) {
        t[i - 1] = t0 + (double) (i - 1) * deltat;
        y[i - 1] = x[0] * pow(t[i - 1], w);
        y[i - 1] += deltay[i - 1] * sigma;
        deltay[i - 1] = sigma;
    }
/* set up A and C */
    i__1 = n;
    for (i = 1; i <= i__1; ++i) {
        c[i - 1] = -y[i - 1];
        a[i - 1] = -pow(t[i - 1], w);
    }
/* perform fit */
    lsqlin_(t, c, deltay, &n, &nr, x, cx, &r, a, scrat, &ok);
/* compute chi**2 probability */
    if (1 == n) {
        p = 0.;
    } else {
        i__1 = n - 1;
        p = 1. - scchi2_(&r, &i__1);
    }
/* prepare graphics */
    memset(tx, ' ', 75);
    memset(ty, ' ', 75);
    strncpy(tx, "t", 1);
    strncpy(ty, "y", 1);
    ltx = 1;
    lty = 1;
    sprintf(capt, "x#=%6.2lf, &D@x=%6.2lf, M=%6.2lf, P=%6.2lf",
            x[0], sqrt(cx[0]), r, p);
    lcapt = strlen(capt);
/* curve corresponding to solution */
    npl = 200;
/* CCC          DT=(DBLE(N-1)/DBLE(NPL-1))*DELTAT */
    dt = (double) n / (double) (npl - 1) * deltat;
    i__1 = npl;
    for (ipl = 1; ipl <= i__1; ++ipl) {
        xpl[ipl - 1] = t0 + dt * (double) ipl;
        ypl[ipl - 1] = x[0] * pow(xpl[ipl - 1], w);
    }
/* prepare data points for graphical presentation */
    i__1 = n;
    for (i = 1; i <= i__1; ++i) {
        datsx[i - 1] = 0.;
        datcov[i - 1] = 0.;
    }
/* graphical output */
    grdtcv_(xpl, ypl, &npl, &c__1, &c_dp5, t, y, datsx, deltay, datcov, &n,
           tx, &ltx, ty, &lty, capt, &lcapt, &nws);
    return 0;
} /* main */

